src/crimson/os/seastore/extentmap_manager/btree/btree_extentmap_manager.cc

```diff
@@ -0,0 +1,119 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+#include <sys/mman.h>
+#include <string.h>
+
+#include "crimson/common/log.h"
+
+#include "include/buffer.h"
+#include "crimson/os/seastore/extentmap_manager/btree/btree_extentmap_manager.h"
+#include "crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore::extentmap_manager {
+
+BtreeExtentMapManager::BtreeExtentMapManager(
+  TransactionManager &tm)
+  : tm(tm) {}
+
+BtreeExtentMapManager::initialize_extmap_ret
+BtreeExtentMapManager::initialize_extmap(Transaction &t)
+{
+
+  logger().debug("BtreeExtentMapManager::initialize_extmap");
+  return tm.alloc_extent<ExtMapLeafNode>(t, L_ADDR_MIN, EXTMAP_BLOCK_SIZE)
+    .safe_then([this](auto&& root_extent) {
+      root_extent->set_size(0);
+      extmap_node_meta_t meta{1};
+      root_extent->set_meta(meta);
+      extmap_root_t extmap_root = extmap_root_t(1, root_extent->get_laddr());
+      return initialize_extmap_ertr::make_ready_future<extmap_root_t>(extmap_root);
+  });
+}
+
+BtreeExtentMapManager::get_root_ret
+BtreeExtentMapManager::get_extmap_root(extmap_root_t &extmap_root, Transaction &t)
+{
+  assert(extmap_root.extmap_root_laddr != L_ADDR_NULL);
+  laddr_t laddr = extmap_root.extmap_root_laddr;
+  return extmap_load_extent(tm, t, laddr, extmap_root.depth);
+}
+
+BtreeExtentMapManager::find_lextent_ret
+BtreeExtentMapManager::find_lextent(extmap_root_t &extmap_root, Transaction &t,
+	                            objaddr_t lo, extent_len_t len)
+{
+  logger().debug("BtreeExtentMapManager::find_lextent: {}, {}", lo, len);
+  return get_extmap_root(extmap_root, t).safe_then([this, &t, lo, len](auto&& extent) {
+    return extent->find_lextent(tm, t, lo, len);
+  }).safe_then([](auto &&e) {
+    logger().debug("BtreeExtentMapManager::find_lextent: found_lextent {}", e);
+    return find_lextent_ret(
+        find_lextent_ertr::ready_future_marker{},
+	std::move(e));
+  });
+
+}
+
+BtreeExtentMapManager::add_lextent_ret
+BtreeExtentMapManager::add_lextent(extmap_root_t &extmap_root, Transaction &t, objaddr_t lo,
+                                   lext_map_val_t val)
+{
+  assert(lo % PAGE_SIZE == 0 && val.length % PAGE_SIZE == 0);
```

Comment:
Does any logic in this module actually care?

Reply:
we have decided extentmap tree only mapping page aligned address and address range, so here add assert to make user use it correctly.

Reply:
This extentmap implementation doesn't care though, it's simply a property we anticipate that the next layer up will maintain.  Nothing here will break if this assert is violated (right?).  I don't see any value in this assert here.

Reply:
assert removed.